pkg/pillar/types/dns.go

```diff
@@ -560,22 +562,32 @@ func GetDNSServers(dns DeviceNetworkStatus, ifname string) []net.IP {
 }
 
 // GetNTPServers returns all, or the ones on one interface if ifname is set
-func GetNTPServers(dns DeviceNetworkStatus, ifname string) []net.IP {
-	var servers []net.IP
+func GetNTPServers(dns DeviceNetworkStatus, ifname string) ([]net.IP, []string) {
+	var serverDomainsOrIPs []string
+	var serverIPs []net.IP
 	for _, us := range dns.Ports {
 		if ifname != "" && ifname != us.IfName {
 			continue
 		}
 		// NTP servers received via DHCP.
-		servers = append(servers, us.NtpServers...)
+		if !us.IgnoreDhcpNtpServers {
+			for _, ntpServer := range us.DhcpNtpServers {
+				// from golang/src/net/ip.go
+				if len(ntpServer) != net.IPv4len && len(ntpServer) != net.IPv6len {
+					logrus.Warnf("parsing ntp server '%v' failed", ntpServer)
+					continue
+				}
+				serverIPs = append(serverIPs, ntpServer)
+			}
+		}
 		// Add statically configured NTP server as well.
-		if us.NtpServer != nil {
-			servers = append(servers, us.NtpServer)
+		if us.ConfiguredNtpServers != nil {
+			serverDomainsOrIPs = append(serverDomainsOrIPs, us.ConfiguredNtpServers...)
 		}
 	}
 	// Avoid duplicates.
-	servers = generics.FilterDuplicatesFn(servers, netutils.EqualIPs)
-	return servers
+	serverDomainsOrIPs = generics.FilterDuplicates(serverDomainsOrIPs)
```

Comment:
Is this FilterDuplicates across a mixture of IP address strings and domain names (which might resolve to some IP addresses already in the list) required? helpful? confusing?

The key filter is down in getIntendedDnsmasqCfg where we make filter out the duplicates based on the resolved IP addresses.

Reply:
Sure, I did.